packages/waku/src/minimal/client.ts

```diff
@@ -0,0 +1,321 @@
+  }
+}
+
+const BASE_PATH = `${import.meta.env?.WAKU_CONFIG_BASE_PATH}${
+  import.meta.env?.WAKU_CONFIG_RSC_BASE
+}/`;
+
+const checkStatus = async (
+  responsePromise: Promise<Response>,
+): Promise<Response> => {
+  const response = await responsePromise;
+  if (!response.ok) {
+    const err = new Error(response.statusText);
+    (err as any).statusCode = response.status;
+    throw err;
+  }
+  return response;
+};
+
+type Elements = Promise<Record<string, ReactNode>> & {
+  prev?: Record<string, ReactNode> | undefined;
+};
+
+const getCached = <T>(c: () => T, m: WeakMap<object, T>, k: object): T =>
+  (m.has(k) ? m : m.set(k, c())).get(k) as T;
+const cache1 = new WeakMap();
+const mergeElements = (a: Elements, b: Elements): Elements => {
+  const getResult = () => {
+    const promise: Elements = new Promise((resolve, reject) => {
+      Promise.all([a, b])
+        .then(([a, b]) => {
+          const nextElements = { ...a, ...b };
+          delete nextElements._value;
+          promise.prev = a;
+          resolve(nextElements);
+        })
+        .catch((e) => {
+          a.then(
+            (a) => {
+              promise.prev = a;
+              reject(e);
+            },
+            () => {
+              promise.prev = a.prev;
+              reject(e);
+            },
+          );
+        });
+    });
+    return promise;
+  };
+  const cache2 = getCached(() => new WeakMap(), cache1, a);
+  return getCached(getResult, cache2, b);
+};
+
+type SetElements = (updater: (prev: Elements) => Elements) => void;
+type EnhanceCreateData = (
+  createData: (
+    responsePromise: Promise<Response>,
+  ) => Promise<Record<string, ReactNode>>,
+) => (responsePromise: Promise<Response>) => Promise<Record<string, ReactNode>>;
+
+const ENTRY = 'e';
+const SET_ELEMENTS = 's';
+const ENHANCE_CREATE_DATA = 'd';
+
+type FetchCache = {
+  [ENTRY]?: [rscPath: string, rscParams: unknown, elements: Elements];
+  [SET_ELEMENTS]?: SetElements;
+  [ENHANCE_CREATE_DATA]?: EnhanceCreateData | undefined;
+};
+
+const defaultFetchCache: FetchCache = {};
+
+/**
+ * callServer callback
+ * This is not a public API.
+ */
+export const callServerRsc = async (
+  funcId: string,
+  args: unknown[],
+  fetchCache = defaultFetchCache,
+) => {
+  const enhanceCreateData = fetchCache[ENHANCE_CREATE_DATA] || ((d) => d);
+  const createData = (responsePromise: Promise<Response>) =>
+    createFromFetch<Awaited<Elements>>(checkStatus(responsePromise), {
+      callServer: (funcId: string, args: unknown[]) =>
+        callServerRsc(funcId, args, fetchCache),
+    });
+  const url = BASE_PATH + encodeRscPath(encodeFuncId(funcId));
+  const responsePromise =
+    args.length === 1 && args[0] instanceof URLSearchParams
+      ? fetch(url + '?' + args[0])
+      : encodeReply(args).then((body) => fetch(url, { method: 'POST', body }));
+  const data = enhanceCreateData(createData)(responsePromise);
+  // FIXME this causes rerenders even if data is empty
```

Comment:
Should this just check for if Object.keys(data).length is zero or is it more complicated?

Reply:
more complicated. `data` here is a promise.